File: minigit-4/ts/src/commands/log.ts
// log command - Show commit history

import { findRepoRoot, formatTimestamp, shortSha } from '../utils';
import { readObject, parseCommitContent, objectExists } from '../objects';
import { getHeadCommit, resolveRevision, getBranches, resolveRef } from '../refs';
import { CommitInfo } from '../types';

export function log(args: string[]): number {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a minigit repository');
    return 1;
  }

  let oneline = false;
  let showAll = false;
  let graph = false;
  let showStat = false;
  let limit: number | null = null;
  let startRef: string | null = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--oneline') {
      oneline = true;
    } else if (args[i] === '--all') {
      showAll = true;
    } else if (args[i] === '--graph') {
      graph = true;
    } else if (args[i] === '--stat') {
      showStat = true;
    } else if (args[i] === '-n' && i + 1 < args.length) {
      limit = parseInt(args[++i], 10);
    } else if (!args[i].startsWith('-')) {
      startRef = args[i];
    }
  }

  // Get starting commits
  let startShas: string[] = [];

  if (showAll) {
    // Get all branch heads
    const branches = getBranches(repoRoot);
    for (const branch of branches) {
      const sha = resolveRef(repoRoot, branch);
      if (sha) startShas.push(sha);
    }
    // Also include HEAD if detached
    const headCommit = getHeadCommit(repoRoot);
    if (headCommit && !startShas.includes(headCommit)) {
      startShas.push(headCommit);
    }
  } else if (startRef) {
    const sha = resolveRevision(repoRoot, startRef);
    if (!sha) {
      console.error(`fatal: bad revision '${startRef}'`);
      return 1;
    }
    startShas = [sha];
  } else {
    const headCommit = getHeadCommit(repoRoot);
    if (!headCommit) {
      console.error('fatal: your current branch does not have any commits yet');
      return 1;
    }
    startShas = [headCommit];
  }

  // Collect commits using BFS (to handle merge commits properly)
  const commits: Array<{ sha: string; info: CommitInfo }> = [];
  const visited = new Set<string>();
  const queue = [...startShas];

  while (queue.length > 0 && (limit === null || commits.length < limit)) {
    // Sort queue by timestamp to show in order
    queue.sort((a, b) => {
      const infoA = getCommitInfo(repoRoot, a);
      const infoB = getCommitInfo(repoRoot, b);
      if (!infoA || !infoB) return 0;
      return infoB.committerTimestamp - infoA.committerTimestamp;
    });

    const sha = queue.shift()!;

    if (visited.has(sha)) continue;
    visited.add(sha);

    if (!objectExists(repoRoot, sha)) continue;

    const { type, content } = readObject(repoRoot, sha);
    if (type !== 'commit') continue;

    const info = parseCommitContent(content);
    commits.push({ sha, info });

    // Add parents to queue
    for (const parent of info.parents) {
      if (!visited.has(parent)) {
        queue.push(parent);
      }
    }
  }

  // Print commits
  for (const { sha, info } of commits) {
    if (oneline) {
      const prefix = graph ? '* ' : '';
      console.log(`${prefix}${shortSha(sha)} ${info.message.split('\n')[0]}`);
    } else {
      if (graph) {
        console.log('*');
      }
      console.log(`commit ${sha}`);
      if (info.parents.length > 1) {
        console.log(`Merge: ${info.parents.map(shortSha).join(' ')}`);
      }
      console.log(`Author: ${info.author} <${info.authorEmail}>`);
      console.log(`Date:   ${formatTimestamp(info.authorTimestamp, info.authorTz)}`);
      console.log('');

      // Indent message
      const messageLines = info.message.split('\n');
      for (const line of messageLines) {
        console.log(`    ${line}`);
      }
      console.log('');

      if (showStat && info.parents.length > 0) {
        // Show stat would require diff calculation
        // For simplicity, just show file names changed
        printCommitStat(repoRoot, sha, info);
      }
    }
  }

  return 0;
}

function getCommitInfo(repoRoot: string, sha: string): CommitInfo | null {
  try {
    const { type, content } = readObject(repoRoot, sha);
    if (type !== 'commit') return null;
    return parseCommitContent(content);
  } catch {
    return null;
  }
}

function printCommitStat(repoRoot: string, sha: string, info: CommitInfo): void {
  if (info.parents.length === 0) return;

  const parentSha = info.parents[0];
  const parentInfo = getCommitInfo(repoRoot, parentSha);
  if (!parentInfo) return;

  // Get files in both trees
  const currentFiles = getTreeFiles(repoRoot, info.tree);
  const parentFiles = getTreeFiles(repoRoot, parentInfo.tree);

  const allFiles = new Set([...currentFiles.keys(), ...parentFiles.keys()]);
  const changedFiles: string[] = [];

  for (const file of allFiles) {
    const currentSha = currentFiles.get(file);
    const parentSha = parentFiles.get(file);

    if (currentSha !== parentSha) {
      changedFiles.push(file);
    }
  }

  for (const file of changedFiles.sort()) {
    console.log(` ${file}`);
  }
  console.log('');
}

function getTreeFiles(repoRoot: string, treeSha: string, prefix: string = ''): Map<string, string> {
  const files = new Map<string, string>();

  try {
    const { content } = readObject(repoRoot, treeSha);
    const { parseTreeContent } = require('../objects');
    const entries = parseTreeContent(content);

    for (const entry of entries) {
      const fullName = prefix ? `${prefix}/${entry.name}` : entry.name;

      if (entry.mode === '40000') {
        const subFiles = getTreeFiles(repoRoot, entry.sha, fullName);
        for (const [name, sha] of subFiles) {
          files.set(name, sha);
        }
      } else {
        files.set(fullName, entry.sha);
      }
    }
  } catch {
    // Ignore errors
  }

  return files;
}

File: minigit-4/py/commands/log.py
"""log command - Show commit history."""

import sys
from pathlib import Path
from datetime import datetime

sys.path.insert(0, str(Path(__file__).parent.parent))

from utils import find_repo_root
from objects import read_object, Commit, resolve_object
from refs import resolve_head, list_branches, get_ref


def format_date(author_line: str) -> str:
    """Extract and format date from author/committer line."""
    # Format: Name <email> timestamp timezone
    parts = author_line.rsplit(' ', 2)
    if len(parts) >= 3:
        try:
            timestamp = int(parts[-2])
            tz = parts[-1]
            dt = datetime.fromtimestamp(timestamp)
            return dt.strftime('%a %b %d %H:%M:%S %Y') + f' {tz}'
        except:
            pass
    return ''


def run(args):
    """Show commit history."""
    repo_root = find_repo_root()
    if repo_root is None:
        print("fatal: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    oneline = False
    show_all = False
    max_count = None
    show_graph = False
    show_stat = False
    start_ref = 'HEAD'

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '--oneline':
            oneline = True
        elif arg == '--all':
            show_all = True
        elif arg == '-n' and i + 1 < len(args):
            max_count = int(args[i + 1])
            i += 1
        elif arg.startswith('-n'):
            max_count = int(arg[2:])
        elif arg == '--graph':
            show_graph = True
        elif arg == '--stat':
            show_stat = True
        elif not arg.startswith('-'):
            start_ref = arg
        i += 1

    # Collect starting points
    start_shas = []
    if show_all:
        for branch in list_branches(repo_root):
            sha = get_ref(f'refs/heads/{branch}', repo_root)
            if sha:
                start_shas.append(sha)
    else:
        try:
            sha = resolve_object(start_ref, repo_root)
            start_shas.append(sha)
        except ValueError as e:
            print(f"fatal: {e}", file=sys.stderr)
            return 1

    if not start_shas:
        print("fatal: no commits to show", file=sys.stderr)
        return 1

    # Walk commit history
    visited = set()
    to_visit = list(start_shas)
    commits = []

    while to_visit:
        sha = to_visit.pop(0)
        if sha in visited:
            continue
        visited.add(sha)

        try:
            obj_type, data = read_object(sha, repo_root)
            if obj_type != 'commit':
                continue

            commit = Commit.deserialize(data)
            commits.append((sha, commit))

            for parent in commit.parents:
                if parent not in visited:
                    to_visit.append(parent)
        except Exception as e:
            continue

    # Sort by timestamp (most recent first)
    def get_timestamp(item):
        sha, commit = item
        # Parse timestamp from author line
        parts = commit.author.rsplit(' ', 2)
        if len(parts) >= 3:
            try:
                return int(parts[-2])
            except:
                pass
        return 0

    commits.sort(key=get_timestamp, reverse=True)

    # Apply max_count
    if max_count is not None:
        commits = commits[:max_count]

    # Output
    for sha, commit in commits:
        if oneline:
            if show_graph:
                print(f'* {sha[:7]} {commit.message.split(chr(10))[0]}')
            else:
                print(f'{sha[:7]} {commit.message.split(chr(10))[0]}')
        else:
            if show_graph:
                print(f'* commit {sha}')
            else:
                print(f'commit {sha}')

            if len(commit.parents) > 1:
                print(f'Merge: {" ".join(p[:7] for p in commit.parents)}')

            # Parse author
            author_parts = commit.author.rsplit(' ', 2)
            if len(author_parts) >= 3:
                author_name = author_parts[0]
            else:
                author_name = commit.author
            print(f'Author: {author_name}')
            print(f'Date:   {format_date(commit.author)}')

            print()
            for line in commit.message.split('\n'):
                print(f'    {line}')
            print()

            if show_stat:
                # Show file changes
                if commit.parents:
                    try:
                        parent_sha = commit.parents[0]
                        _, parent_data = read_object(parent_sha, repo_root)
                        parent_commit = Commit.deserialize(parent_data)
                        parent_files = get_tree_files(parent_commit.tree_sha, repo_root)
                    except:
                        parent_files = {}
                else:
                    parent_files = {}

                current_files = get_tree_files(commit.tree_sha, repo_root)

                # Find differences
                all_files = set(parent_files.keys()) | set(current_files.keys())
                for f in sorted(all_files):
                    if f not in parent_files:
                        print(f' {f} | new file')
                    elif f not in current_files:
                        print(f' {f} | deleted')
                    elif parent_files[f] != current_files[f]:
                        print(f' {f} | modified')
                print()

    return 0


def get_tree_files(tree_sha: str, repo_root: Path) -> dict:
    """Get all files from a tree."""
    from objects import Tree

    files = {}

    def walk_tree(sha: str, prefix: str):
        try:
            _, data = read_object(sha, repo_root)
            tree = Tree.deserialize(data)
            for entry in tree.entries:
                path = f'{prefix}{entry.name}' if prefix else entry.name
                if entry.mode.startswith('40'):
                    walk_tree(entry.sha, path + '/')
                else:
                    files[path] = entry.sha
        except:
            pass

    walk_tree(tree_sha, '')
    return files
